[Merged by Bors] - feat: path-connectedness is preserved by homeomorphisms

grunweg opened the pull request:
Mirrors the corresponding proofs for connectedness.

Co-authored-by: @ADedecker

grunweg commented:
Removing easy label as this PR adds two simp lemmas (even though they're probably uncontroversial). Sorry for the churn.

ADedecker commented on Mathlib/Topology/Connected/PathConnected.lean:
```diff
@@ -969,10 +970,25 @@ theorem IsPathConnected.image' {Y : Type*} [TopologicalSpace Y] (hF : IsPathConn
 #align is_path_connected.image IsPathConnected.image
 
+/-- If `h : X → Y` is a homeomorphism, `h(s)` is path-connected iff `s` is. -/
+-- this is not in `Homeomorph.lean` to avoid import cycles
```
```suggestion
```
I don't think this is a controversial statement, so I'd remove this line. In fact, I would argue that `Homeomorph.isConnected_image` should be in the connectedness file, but that's not worth refactoring.

grunweg replied:
Sure, removed.

ADedecker reviewed (commented):
LGTM!

If you're interested, I made a `Inducing` version. You'd need to add:
- in `UnitInterval.lean`: 
```lean
instance : NeZero (1 : I) := ⟨fun h ↦ one_ne_zero <| congrArg Subtype.val h⟩
```
- in `PathConnected.lean`:
```lean
theorem Specializes.joinedIn (h : x ⤳ y) (hx : x ∈ F) (hy : y ∈ F) : JoinedIn F x y := by
  refine ⟨⟨⟨Set.piecewise {1} (const I y) (const I x), ?_⟩, by simp, by simp⟩, fun t ↦ ?_⟩
  · simp only [const, continuous_def, piecewise_preimage]
    intro U hU
    by_cases hy' : y ∈ U
    · simp [hy', h.mem_open hU hy']
    · by_cases hx' : x ∈ U
      · simpa [hx', hy'] using isOpen_univ.sdiff isClosed_singleton
      · simp [hx', hy']
  · simp only [Path.coe_mk_mk, piecewise]
    split_ifs <;> assumption

theorem Inseparable.joinedIn (h : Inseparable x y) (hx : x ∈ F) (hy : y ∈ F) : JoinedIn F x y :=
  h.specializes.joinedIn hx hy
```
And then:
```lean
nonrec theorem Inducing.isPathConnected_iff {Y : Type*} [TopologicalSpace Y]
    {f : X → Y} (hf : Inducing f) :
    IsPathConnected F ↔ IsPathConnected (f '' F) := by
  refine ⟨fun hF ↦ hF.image hf.continuous, fun hF ↦ ?_⟩
  simp? [isPathConnected_iff] at hF ⊢ says
    simp only [isPathConnected_iff, nonempty_image_iff, mem_image, forall_exists_index,
      and_imp, forall_apply_eq_imp_iff₂] at hF ⊢
  refine ⟨hF.1, fun x hx y hy ↦ ?_⟩
  rcases hF.2 x hx y hy with ⟨γ, hγ⟩
  choose γ' hγ' hγγ' using hγ
  have key₁ : Inseparable x (γ' 0) := by rw [← hf.inseparable_iff, hγγ' 0, γ.source]
  have key₂ : Inseparable (γ' 1) y := by rw [← hf.inseparable_iff, hγγ' 1, γ.target]
  refine key₁.joinedIn hx (hγ' 0) |>.trans ⟨⟨⟨γ', ?_⟩, rfl, rfl⟩, hγ'⟩ |>.trans
    (key₂.joinedIn (hγ' 1) hy)
  simpa [hf.continuous_iff] using γ.continuous.congr fun t ↦ (hγγ' t).symm
```

If you prefer I can do that in a later PR.

grunweg commented:
@ADedecker Wow! I didn't mean to nerd-snipe you - but now that you've suggested this, I have added it. I don't know how to add you as a co-author so git recognises this; feel free to force-push something/edit the top-level comment accordingly.

I saw that the file already had the implicit variables you mentioned, hence went over the file and removed all superfluous `{X : Type*} {TopologicalSpace X}` (also so Y). Let me know if I should drop that commit/move to a different PR.

ADedecker commented:
Don't worry for nerd-sniping, that was very fun! I've added my name to the PR description. This looks good to me, but obviously I'm not going to merge it myself, I'll ask on Zulip.

jcommelin commented:
Thanks :tada:

bors merge
